onnxruntime/core/providers/openvino/qdq_transformations/qdq_stripping.cc

```diff
@@ -0,0 +1,699 @@
+// Copyright (C) Intel Corporation
+// Licensed under the MIT License
+
+#include <array>
+#include <vector>
+#include <iostream>
+#include <fstream>
+#include <algorithm>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "core/providers/shared_library/provider_api.h"
+#include "core/providers/openvino/qdq_transformations/qdq_stripping.h"
+
+namespace onnxruntime {
+namespace openvino_ep {
+
+enum class SkipReason {
+  Uint16QDQ,
+  DuplicateDQ,
+  ConstInitUnsupportedDQ,
+  SandwichedDQ,
+  Other
+};
+
+constexpr std::string_view DuplicateDQ = "/duplicated";
+
+// Return the data type of the qdq node.
+// Check output type of Q and input type of DQ to determine it as zero_point is an optional input and may not exist
+static ONNX_NAMESPACE::TensorProto_DataType GetQDQDataType(const Node* qdq_node) {
+  if (qdq_node->OpType() == "QuantizeLinear") {
+    return static_cast<ONNX_NAMESPACE::TensorProto_DataType>(
+        qdq_node->OutputDefs().at(0)->TypeAsProto()->tensor_type().elem_type());
+  } else if (qdq_node->OpType() == "DequantizeLinear") {
+    return static_cast<ONNX_NAMESPACE::TensorProto_DataType>(
+        qdq_node->InputDefs().at(0)->TypeAsProto()->tensor_type().elem_type());
+  } else {
+    ORT_THROW("Invalid QDQ Op Type when fetching datatype of parameter");
+  }
+}
+
+// Creates a new NodeArg from an input (or output) of a QDQ node unit. If the input/output is quantized,
+// this function modifies the tensor type to the specified float type.
+static NodeArg& ProcessNodeUnitIO(onnxruntime::Graph& dst_graph,
+                                  const onnxruntime::GraphViewer& src_graph,
+                                  std::set<std::string>& initializers_to_keep,
+                                  const NodeUnitIODef& io_def) {
+  const std::string& name = io_def.node_arg.Name();
+  const ONNX_NAMESPACE::TypeProto* orig_type_proto = io_def.node_arg.TypeAsProto();
+
+  // Handle quantized input or output. Convert to float type.
+  if (io_def.quant_param.has_value()) {
+    // Copy the original quantized type proto, but update the type to float.
+    std::unique_ptr<ONNX_NAMESPACE::TypeProto> type_proto = ONNX_NAMESPACE::TypeProto::Create();
+    type_proto->copy_from(orig_type_proto);
+    type_proto->mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
```

Comment:
You could get the actual float type from the type of the scale in the Q (or DQ) op.

One way to do this is to get scale initializer's name -> get initializer TensorProto -> get the float type
```c++
const auto& src_initializers = src_graph.GetAllInitizedTensors();
const std::string& scale_initializer_name = io_def.quant_param->scale.Name();
auto tensor_proto_iter = src_initializers.find(scale_initializer_name);

// Should check that it exists. Maybe change signature of this function to return a Status.
// ORT_RETURN_IF(tensor_proto_iter == src_initializers.end(), "Unable to find scale initializer ", scale_initializer_name);

const ONNX_NAMESPACE::TensorProto* scale_tensor_proto = tensor_proto_iter->second;
int32_t float_type = scale_tensor_proto->data_type();
```

Reply:
Fixed